Improve Guidebook *roff formatting and markup

g-branden-robinson opened the pull request:
Turning on _groff_ warnings can help document maintainers spot and correct bugs.  Get `Guidebook.mn` (and `tmac.nh`) into good shape for this purpose.

g-branden-robinson committed:
doc/Guidebook.mn: Fix table markup nits.

Comment escape sequences inside table entries can wreak havoc.  Use
dummy character instead to visually indicate the deliberate trailing
spaces.  Move the comment explaining why they're there closer to what
they document.  It's okay to have _whole-line_ comments in table data,
because they are on control lines (lines that start with a dot '.').

Also use the dummy character to indicate deliberately empty table cells.

g-branden-robinson committed:
doc/Guidebook.mn: Simplify table format.

The 'A' ("alphabetic") and 'N' ("numeric") column classifiers were being
used to little benefit.

Since 'A' was applying to every row of the table, none was more indented
than any other, except via the inclusion of unadjustable, unbreakable
space escape sequences `\ `, which work just as well with column
classifier 'L' ("left").

In fact, even they are unnecessary; regular spaces will do.

tbl(1):
     Ordinarily, a table entry is typeset rigidly.  It is not filled,
     broken, hyphenated, adjusted, or populated with additional inter‐
     sentence space.

...so furthermore convert the escaped spaces to regular ones.

Similarly, 'N' applies several rules to manage alignment of decimal
points.  This table doesn't need them.  Right-alignment of integers is
just as easily achieved with the 'R' ("right") column classifier.

g-branden-robinson committed:
doc/Guidebook.mn: Refactor achievement table.

Use tbl(1); this sort of layout is what it's designed for.  Use text
blocks to enable lengthy achievement descriptions to break across lines.

g-branden-robinson committed:
doc/Guidebook.mn: Fix typo in escape sequence.

\F and \f do different things.

Fixes:

$ (cd doc && cat Guidebook.mn | tbl tmac.n - | groff > Guidebook.ps)
troff:<standard input>:3468: error: no font family named 'I' exists

g-branden-robinson committed:
doc/tmac.n: Fix erroneous lines.

I'm mindful of the license here, but suspect that these lines crept in
after Matt Bishop's time.  Text lines in a macro package can be
insidious because they cause formatting operations to start even in the
absence of an input document.  The forthcoming groff 1.24 has a new
diagnostic to help catch these situations.

The groff Texinfo manual says:

     A '\"' comment on a line by itself is treated as a blank line,
     because after eliminating the comment, that is all that remains.

          Test
          \" comment
          Test
              => Test
              =>
              => Test

     To compensate, it is common to combine the empty request with the
     comment escape sequence as '.\"', causing the input line to be
     ignored.

Fixes:

$ (cd doc && groff -t -M . -mn -mnh Guidebook.mn > /dev/null)
troffrc:./tmac.n:1: text line in startup file
troffrc:./tmac.n:764: text line in startup file

g-branden-robinson committed:
Make it easier for the Guidebook to find macros.

* doc/Guidebook.mn: Make the Guidebook buildable from the top of the
  source tree, not just inside the "doc" directory.  Try to load its
  "nh" macro package from the current working directory and from "doc".

* doc/tmac.nh: Allocate new register `nH` to the purpose of detecting
  multiple loads, and skip file content if detected.  This is the 1970s
  nroff form of an "#include guard".  groff's "an-ext.tmac" uses the
  same technique for portability.

Also I removed a tab character.  Per the groff Texinfo manual:

     One possibly irritating idiosyncrasy is that tabs should not be
     used to vertically align comments in the source document.  Tab
     characters are not treated as separators between a request name and
     its first argument, nor between arguments.

Here's an example of how one groff macro package works around the
problem.

$ sed -n '402,406p' contrib/mm/m.tmac
.ds LetCN CONFIDENTIAL\"                Confidential default
.ds LetSA To Whom It May Concern:\"     Salutation default
.ds LetAT ATTENTION:\"                  Attention string
.ds LetSJ SUBJECT:\"                    Subject string
.ds LetRN In reference to:\"            Reference string

g-branden-robinson committed:
doc/Guidebook.mn: Use different technique.

...to get a block paragraph from the `n` package, instead of
manipulating indentation directly, which can lead to grief.

g-branden-robinson committed:
doc/Guidebook.mn: Fix bad comment syntax.

These lines were passing a spurious "." argument to the `BR` macro
defined in "doc/tmac.nh".

g-branden-robinson committed:
doc/tmac.nh: Fix mistaken syntax.

The forthcoming groff 1.24 has a new diagnostic that detects ill-formed
numeric expressions.  It has found one here.

I'm not positive what was intended here, but it may have been an attempt
to force interpretation of the first macro argument as a number.  This
change employs a more idiomatic (but still old-school) technique.

The salient fact is that, in *roff, you can't affix a scaling unit after
a closing parenthesis (or another scaling unit).

In GNU troff the `\B` escape sequence, an extension, permits the testing
of putative numeric expressions for validity.

troff:./doc/Guidebook.mn:268: warning: expected end of line or an
auto-increment argument in register definition request; got character
'v'

See <https://savannah.gnu.org/bugs/?64240>.

g-branden-robinson committed:
doc/tmac.nh (ED): Don't dedent below zero.

This quietens several warnings from GNU troff in the "range" category.

troff: backtrace: './doc/tmac.nh':92: macro 'ED'
troff: backtrace: file './doc/Guidebook.mn':5950
troff:./doc/Guidebook.mn:5950: warning: treating -120u indentation as
zero

Unfortunately, the similar `ed` macro in Bishop's "mn" package
contributes several more.

But with this change (and its forerunners in this series), the NetHack
Guidebook is now warning-free with "-wall -Wtab -Wrange" ("mn" has
problems with tab characters too), even with the increasingly fastidious
syntactical checks of the forthcoming groff 1.24 release.

g-branden-robinson committed:
Build *roff documents with groff warnings enabled.

...some of them, at any rate.  We shut off (1) warnings provoked by Matt
Bishop's "mn" macro package, and (2) warnings spuriously emitted by
groff 1.23 and previous when "-wall" (or "-ww") is specified.

Also update explanatory comments.

nhmall commented:
Here are the results from a test with this pull request merged into branch NetHack-3.7:
```
make Guidebook
cat Guidebook.dated.mn | ../util/makedefs --grep --input - --output - | tbl tmac.n - | nroff -c -Tascii -wall -Wtab -Wrange | col -bx > Guidebook
troff:tmac.n:732: warning: 'p' scaling unit invalid in context; convert to 'z' or 'u'
troff:tmac.n:733: warning: 'p' scaling unit invalid in context; convert to 'z' or 'u'
troff:tmac.n:762: error: cannot load font 'S' for emboldening
troff:<standard input>:67: warning: 'p' scaling unit invalid in context; convert to 'z' or 'u'
troff:<standard input>:75: warning: unbalanced 'el' request
troff:<standard input>:75: warning: unbalanced 'el' request
troff:<standard input>:122: warning: unbalanced 'el' request
troff:<standard input>:122: warning: unbalanced 'el' request
troff:<standard input>:211: warning: unbalanced 'el' request
troff:<standard input>:211: warning: unbalanced 'el' request
troff:<standard input>:307: warning: unbalanced 'el' request
troff:<standard input>:442: warning: unbalanced 'el' request
troff:<standard input>:457: warning: unbalanced 'el' request
troff:<standard input>:549: warning: unbalanced 'el' request
troff:<standard input>:549: warning: unbalanced 'el' request
troff:<standard input>:2023: warning: unbalanced 'el' request
troff:<standard input>:2023: warning: unbalanced 'el' request
troff:<standard input>:2039: warning: unbalanced 'el' request
troff:<standard input>:2100: warning: unbalanced 'el' request
troff:<standard input>:2171: warning: unbalanced 'el' request
troff:<standard input>:2207: warning: unbalanced 'el' request
troff:<standard input>:2261: warning: unbalanced 'el' request
troff:<standard input>:2339: warning: unbalanced 'el' request
troff:<standard input>:2357: warning: unbalanced 'el' request
troff:<standard input>:2357: warning: unbalanced 'el' request
troff:<standard input>:2377: warning: unbalanced 'el' request
troff:<standard input>:2404: warning: unbalanced 'el' request
troff:<standard input>:2429: warning: unbalanced 'el' request
troff:<standard input>:2454: warning: unbalanced 'el' request
troff:<standard input>:2465: warning: unbalanced 'el' request
troff:<standard input>:2483: warning: unbalanced 'el' request
troff:<standard input>:2483: warning: unbalanced 'el' request
troff:<standard input>:2536: warning: unbalanced 'el' request
troff:<standard input>:2583: warning: unbalanced 'el' request
troff:<standard input>:2809: warning: unbalanced 'el' request
troff:<standard input>:2885: warning: unbalanced 'el' request
troff:<standard input>:2912: warning: unbalanced 'el' request
troff:<standard input>:2942: warning: unbalanced 'el' request
troff:<standard input>:2956: warning: unbalanced 'el' request
troff:<standard input>:2989: warning: unbalanced 'el' request
troff:<standard input>:3012: warning: unbalanced 'el' request
troff:<standard input>:3067: warning: unbalanced 'el' request
troff:<standard input>:3133: warning: unbalanced 'el' request
troff:<standard input>:3151: warning: unbalanced 'el' request
troff:<standard input>:3162: warning: unbalanced 'el' request
troff:<standard input>:3191: warning: unbalanced 'el' request
troff:<standard input>:3212: warning: unbalanced 'el' request
troff:<standard input>:3237: warning: unbalanced 'el' request
troff:<standard input>:3237: warning: unbalanced 'el' request
troff:<standard input>:3385: warning: unbalanced 'el' request
troff:<standard input>:3477: warning: unbalanced 'el' request
troff:<standard input>:3477: warning: unbalanced 'el' request
troff:<standard input>:3483: warning: unbalanced 'el' request
troff:<standard input>:3494: warning: unbalanced 'el' request
troff:<standard input>:3702: warning: unbalanced 'el' request
troff:<standard input>:3752: warning: unbalanced 'el' request
troff:<standard input>:4823: warning: unbalanced 'el' request
troff:<standard input>:5038: warning: unbalanced 'el' request
troff:<standard input>:5054: warning: unbalanced 'el' request
troff:<standard input>:5126: warning: unbalanced 'el' request
troff:<standard input>:5137: warning: unbalanced 'el' request
troff:<standard input>:5197: warning: unbalanced 'el' request
troff:<standard input>:5345: warning: unbalanced 'el' request
troff:<standard input>:5395: warning: unbalanced 'el' request
troff:<standard input>:5455: warning: unbalanced 'el' request
troff:<standard input>:5498: warning: unbalanced 'el' request
troff:<standard input>:5682: warning: unbalanced 'el' request
troff:<standard input>:5925: warning: unbalanced 'el' request
troff:<standard input>:5962: warning: unbalanced 'el' request
troff:<standard input>:6038: warning: unbalanced 'el' request
troff:<standard input>:6173: warning: unbalanced 'el' request
troff:<standard input>:6173: warning: unbalanced 'el' request
troff:<standard input>:6197: warning: unbalanced 'el' request
troff:<standard input>:6197: warning: unbalanced 'el' request
troff:<standard input>:6222: warning: unbalanced 'el' request
troff:<standard input>:6245: warning: unbalanced 'el' request
troff:<standard input>:6245: warning: unbalanced 'el' request
troff:<standard input>:6592: warning: unbalanced 'el' request
troff:<standard input>:6608: warning: unbalanced 'el' request
```
The results of `../util/makedefs --grep --input - --output - | tbl tmac.n - >out1.txt` can be found here:
https://gist.github.com/nhmall/6739cd96c526f6b2399c49d842c41012

nroff --version
GNU nroff (groff) version 1.23.0
GNU groff version 1.23.0
Copyright (C) 2022 Free Software Foundation, Inc.
GNU groff comes with ABSOLUTELY NO WARRANTY.
You may redistribute copies of groff and its subprograms
under the terms of the GNU General Public License.
For more information about these matters, see the file
named COPYING.

called subprograms:

GNU grotty (groff) version 1.23.0
GNU troff (groff) version 1.23.0

g-branden-robinson commented:
> Here are the results from a test with this pull request merged into branch NetHack-3.7:

Thanks for testing these changes with the current stable release of _groff_.  I get the same when I build the Guidebook with 1.23.0.
 
> ```
> make Guidebook
> cat Guidebook.dated.mn | ../util/makedefs --grep --input - --output - | tbl tmac.n - | nroff -c -Tascii -wall -Wtab -Wrange | col -bx > Guidebook

[condensing to unique diagnostic message texts]


> troff:tmac.n:732: warning: 'p' scaling unit invalid in context; convert to 'z' or 'u'

That is [Savannah #66150](https://savannah.gnu.org/bugs/?66150), fixed in Git on 4 September.  (In fact, observing the historical `mn` macro package's use of an explicit `p` scaling unit is what motivated me to fix that _groff_ bug!)

> troff:tmac.n:762: error: cannot load font 'S' for emboldening

That is [Savannah #64166](https://savannah.gnu.org/bugs/?64166), fixed in Git on 10 July 2023.

> troff:<standard input>:75: warning: unbalanced 'el' request

That is [Savannah #65474](https://savannah.gnu.org/bugs/?65474), fixed in Git on 5 April.

g-branden-robinson commented:
I've updated the final commit in the series.

It:

1. Quietens the riot of warnings from _groff_ 1.23.0.
2. Prepares for the future release of _groff_ 1.24.0.

With these changes I get only one error message each (the first of which is already extant in the NetHack-3.7 branch).

```
$ ~/groff-stable/bin/groff --version | tail -n 1
GNU troff (groff) version 1.23.0
$ rm Guidebook
$ make Guidebook
cat Guidebook.dated.mn | ../util/makedefs --grep --input - --output - | tbl tmac.n - | nroff -c -Tascii -wall -Wtab -Wrange -Wel -Wscale | col -bx > Guidebook
troff:tmac.n:762: error: cannot load font 'S' for emboldening

$ groff --version | tail -n 1
GNU troff (groff) version 1.23.0.2006-a00de
$ rm Guidebook
$ make Guidebook
cat Guidebook.dated.mn | ../util/makedefs --grep --input - --output - | tbl tmac.n - | nroff -c -Tascii -wall -Wtab -Wrange -Wel -Wscale | col -bx > Guidebook
troff: error: unrecognized warning category 'el'
```

The second error will not arise when building the Guidebook with a released _groff_ 1.24; as is shown, the formatter is still advertising a version number of 1.23.  So only I, and people bold enough to run _groff_ from Git, will ever see it.
